paddle/fluid/operators/mkldnn/mul_mkldnn_op.cc

```diff
@@ -0,0 +1,497 @@
+        this->CreateMemDescriptor(&x_matrix, x_dt, memory::format::nc);
+    this->input_ = this->CreateMemory(src_desc, &x_matrix);
+
+    if (std::is_same<K, float>::value) {
+      auto weights_dt = framework::ToMKLDNNDataType(
+          framework::DataTypeTrait<float>::DataType);
+      auto trans_weights_ = this->TransposeWeights(&y_matrix, weights_dt);
+      this->weights_ = this->UpdateWeights(trans_weights_, scale_weight);
+    } else if (std::is_same<K, int8_t>::value) {
+      this->weights_ = this->TransposeWeights(
+          &y_matrix, framework::ToMKLDNNDataType(
+                         framework::DataTypeTrait<int8_t>::DataType));
+    }
+
+    bool force_fp32_output = ctx.Attr<bool>("force_fp32_output");
+    memory::data_type dst_dt;
+    if (force_fp32_output) {
+      dst_dt = framework::ToMKLDNNDataType(
+          framework::DataTypeTrait<float>::DataType);
+    } else {
+      dst_dt = framework::ToMKLDNNDataType(
+          framework::DataTypeTrait<int8_t>::DataType);
+    }
+
+    auto dst_desc =
+        this->CreateMemDescriptor(output, dst_dt, memory::format::any);
+
+    this->fc_ = this->CreateFcPrimitive(*(this->input_), *(this->weights_),
+                                        dst_desc, output, ctx);
+    return *(this->fc_);
+  }
+
+  virtual mkldnn::memory Reorder(const memory::desc& src_desc,
+                                 const memory::desc& dst_desc,
+                                 const void* src_data,
+                                 const std::vector<float>& scale = {1.0f}) {
+    auto mask = scale.size() > 1 ? 1 : 0;
+    mkldnn::primitive_attr attr;
+    attr.set_output_scales(mask, scale);
+    auto src_mem =
+        memory({src_desc, this->engine_}, const_cast<void*>(src_data));
+    auto dst_mem = memory({dst_desc, this->engine_});
+    auto reorder_pd = mkldnn::reorder::primitive_desc(
+        src_mem.get_primitive_desc(), dst_mem.get_primitive_desc(), attr);
+
+    auto reorder = mkldnn::reorder(reorder_pd, src_mem, dst_mem);
+    stream(stream::kind::eager).submit({reorder}).wait();
+
+    return dst_mem;
+  }
+
+  mkldnn::memory TransposeWeights(const Tensor* weights,
+                                  mkldnn::memory::data_type type) {
+    auto dims = framework::vectorize2int(weights->dims());
+    std::swap(dims[0], dims[1]);  // Correct output dimensions
+    auto src_desc = this->CreateMemDescriptor(dims, type, memory::format::io);
+    auto dst_desc = this->CreateMemDescriptor(dims, type, memory::format::oi);
+    // MOD: return Reorder(src_desc, dst_desc, weights->data<T>());
+    return Reorder(src_desc, dst_desc, weights->data<K>());
+  }
+
+  mkldnn::memory UpdateWeights(mkldnn::memory weights,
+                               const std::vector<float>& scale_weight) {
+    auto dims = weights.get_primitive_desc().desc().data.dims;
+    auto ndims = weights.get_primitive_desc().desc().data.ndims;
+    auto weights_dims = std::vector<int>(dims, dims + ndims);
```

Comment:
Wouldn't it be sufficient here to just make a copy of weights desc and update the datatype of the destination memory descriptor ?

Reply:
okay, i will try to simplify the implementation of this function.

Reply:
I tried to simplify the code, it results in an exception while running the unit test,  tried to figure it out the reason but still don't know why for yet, since the deadline is coming, so I kept the original code for now.

Reply:
done